common/models/user.js

```diff
@@ -580,7 +580,10 @@ module.exports = function(User) {
             query && query.redirect !== undefined ? query.redirect :
             undefined;
 
-          ctx.res.redirect(redirectUrl);
+          if (redirectUrl !== undefined) {
+            ctx.res.location(redirectUrl);
+          }
+          next()
         } else {
           next(new Error('transport unsupported'));
```

Comment:
If there is a `redirectUrl` parameter, then the caller of the API is expecting to receive a redirect. Since non-REST transports (JSON-RPC, WebSocket) don't support redirects, an error should be reported in such case, which is the current behaviour.